duecredit/io.py

```diff
@@ -72,93 +72,96 @@ def dump(self, tags=None):
+            packages[key] = defaultdict(list)
+            modules[key] = defaultdict(list)
+            objects[key] = defaultdict(list)
+
+        # for each path store both a list of entry keys and of citations
+        for (path, entry_key), citation in iteritems(citations):
+            if ':' in path:
+                target_dict = objects
+            elif '.' in path:
+                target_dict = modules
+            else:
+                target_dict = packages
+            target_dict['citations'][path].append(citation)
+            target_dict['entry_keys'][path].append(entry_key)
+        return packages, modules, objects
 
-        # Separate logic (model) from presentation (view).  Let's first create a "model"
-        # Collect all citations under their corresponding packages
-
-        # TODO: such logic/setup would not work if we want to allow citations for modules
-        # within packages, so we really need a 3 level reporting:  package / module / obj
-        cited_packages = {}
-        for citation in itervalues(citations):
-            package = citation.package
-            objname = citation.objname
-
-            if package not in cited_packages:
-                # list of two lists -- one citations for the package itself,
-                # another one will be also dictionary for citations for functions
-                cited_packages[package] = [[], {}]
+    def dump(self, tags=None):
+        # get 'model' of citations
+        packages, modules, objects = self._model_citations(tags)
 
-            if citation.cites_module is True:
-                cited_packages[package][0].append(citation)
-            else:
-                if objname not in cited_packages[package][1]:
-                    # initiate a list of citations for that object
-                    cited_packages[package][1][objname] = []
-                cited_packages[package][1][objname].append(citation)
-
-        # Now prune references to packages which had no citations ot internal functionality
-        # TODO: theoretically should be done before pruning based on tags so we still
-        # catch those which were used anyhow
-        for package, (package_citations, obj_citations) in list(iteritems(cited_packages)): # operate on a copy
-            # check if any module citation is "forced", so we
-            # always cite if it was imported
-            if any(c.cite_module for c in package_citations):
-                continue
-            if not obj_citations:
-                cited_packages.pop(package)
+        citations_ordered = []
+        # set up view
+        keys2refnr = defaultdict(int)  # mapping key -> ref nr
+        refnr = 1
+
+        # package level
+        for package in sorted(packages['entry_keys']):
+            for entry_key in packages['entry_keys'][package]:
+                if entry_key not in keys2refnr:
+                    keys2refnr[entry_key] = refnr
+                    refnr += 1
```

Comment:
that is what I felt -- keys2refnr shouldn't be just a dict but rather a class which might inside use a dict but this 3 line functionality should pretty much inside, not duplicated around, and then may be other stuff would become obvious how to refactor, e.g.
```python
class EntriesEnumerated(object):
   ...
   def __getitem__(self, entry_key):
               if entry_key not in self._keys2refnr:
                    self._keys2refnr[entry_key] = self._refnr
                    self._refnr += 1
              return self._keys2refnr[entry_key]
```